core/types/types.go

```diff
@@ -111,19 +152,303 @@ func (bh *Header) Hash() common.Hash {
+		enc = append(enc, venc...)
+	} else {
+		enc = append(enc, byte(0)) // None
+	}
+
+	if bd.Receipt.Exists() {
+		venc, err := scale.Encode(bd.Receipt.Value())
+		if err != nil {
+			return nil, err
+		}
+		enc = append(enc, byte(1)) // Some
+		enc = append(enc, venc...)
+	} else {
+		enc = append(enc, byte(0)) // None
+	}
+
+	if bd.MessageQueue.Exists() {
+		venc, err := scale.Encode(bd.MessageQueue.Value())
+		if err != nil {
+			return nil, err
+		}
+		enc = append(enc, byte(1)) // Some
+		enc = append(enc, venc...)
+	} else {
+		enc = append(enc, byte(0)) // None
+	}
+
+	if bd.Justification.Exists() {
+		venc, err := scale.Encode(bd.Justification.Value())
+		if err != nil {
+			return nil, err
+		}
+		enc = append(enc, byte(1)) // Some
+		enc = append(enc, venc...)
+	} else {
+		enc = append(enc, byte(0)) // None
+	}
+
+	return enc, nil
+}
+
+// Decode decodes the SCALE encoded input to BlockData
+func (bd *BlockData) Decode(r io.Reader) error {
+	hash, err := common.ReadHash(r)
+	if err != nil {
+		return err
+	}
+	bd.Hash = hash
+
+	bd.Header, err = decodeOptionalHeader(r)
+	if err != nil {
+		return err
+	}
+
+	bd.Body, err = decodeOptionalBody(r)
+	if err != nil {
+		return err
+	}
+
+	bd.Receipt, err = decodeOptionalBytes(r)
+	if err != nil {
+		return err
+	}
+
+	bd.MessageQueue, err = decodeOptionalBytes(r)
+	if err != nil {
+		return err
+	}
+
+	bd.Justification, err = decodeOptionalBytes(r)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// EncodeBlockDataArray encodes an array of BlockData using SCALE
+func EncodeBlockDataArray(bds []*BlockData) ([]byte, error) {
+	enc, err := scale.Encode(int32(len(bds)))
+	if err != nil {
+		return nil, err
+	}
+
+	for _, bd := range bds {
+		benc, err := bd.Encode()
+		if err != nil {
+			return nil, err
+		}
+		enc = append(enc, benc...)
+	}
+
+	return enc, nil
+}
+
+// DecodeBlockDataArray decodes a SCALE encoded BlockData array
+func DecodeBlockDataArray(r io.Reader) ([]*BlockData, error) {
+	sd := scale.Decoder{Reader: r}
+
+	l, err := sd.Decode(int32(0))
+	if err != nil {
+		return nil, err
+	}
+
+	length := int(l.(int32))
+	bds := make([]*BlockData, length)
+
+	for i := 0; i < length; i++ {
+		bd := new(BlockData)
+		err = bd.Decode(r)
+		if err != nil {
+			return bds, err
+		}
+
+		bds[i] = bd
+	}
+
+	return bds, err
+}
+
+// decodeOptionalHeader decodes a SCALE encoded optional Header into an *optional.Header
+func decodeOptionalHeader(r io.Reader) (*optional.Header, error) {
+	sd := scale.Decoder{Reader: r}
+
+	exists, err := common.ReadByte(r)
+	if err != nil {
+		return nil, err
+	}
+
+	if exists == 1 {
+		header := &Header{
+			ParentHash:     common.Hash{},
+			Number:         big.NewInt(0),
+			StateRoot:      common.Hash{},
+			ExtrinsicsRoot: common.Hash{},
+			Digest:         [][]byte{}, // TODO: digests are not decoded properly due to SCALE
+		}
+		_, err = sd.Decode(header)
+		if err != nil {
+			return nil, err
+		}
+
+		header.Hash()
+
+		// TODO: fix SCALE :(
```

Comment:
#584

Reply:
I can fix this up if #587 gets merged before this